mem/buffer_slice.go

```diff
@@ -224,3 +228,47 @@ func (w *writer) Write(p []byte) (n int, err error) {
 func NewWriter(buffers *BufferSlice, pool BufferPool) io.Writer {
 	return &writer{buffers: buffers, pool: pool}
 }
+
+// ReadAll reads from r until an error or EOF and returns the data it read.
+// A successful call returns err == nil, not err == EOF. Because ReadAll is
+// defined to read from src until EOF, it does not treat an EOF from Read
+// as an error to be reported.
+// Make sure to free the returned buffers even if you get an error from this function.
+func ReadAll(r io.Reader, pool BufferPool) (BufferSlice, error) {
+	var result BufferSlice
+	wt, ok := r.(io.WriterTo)
+	if ok {
+		// This is more optimal since wt knows the size of chunks it wants to write and, hence, we can allocate
+		// buffers of an optimal size to fit them. E.g. might be a single big chunk, and we wouldn't chop it into pieces.
+		w := NewWriter(&result, pool)
+		_, err := wt.WriteTo(w)
+		return result, err
+	}
+	for {
+		buf := pool.Get(readAllBufSize)
+		// We asked for 32KiB but may have been given a bigger buffer. Use all of it if that's the case.
+		*buf = (*buf)[:cap(*buf)]
+		usedCap := 0
+		for {
+			n, err := r.Read((*buf)[usedCap:])
+			usedCap += n
+			if err != nil {
+				if usedCap == 0 {
+					// Nothing in this buf, put it back
+					pool.Put(buf)
+				} else {
+					*buf = (*buf)[:usedCap]
+					result = append(result, NewBuffer(buf, pool))
+				}
```

Comment:
Unless there's an explicit usecase where a caller of `ReadAll` would want to have access to the data on a partial read, it's probably better to have a saner default where the data is freed. This is a convention that's explicitly called out [in the docs for BufferSlice](https://github.com/grpc/grpc-go/blob/master/mem/buffer_slice.go#L36). If you want to break from the convention because there is a hard need for it, that's fine but you'd want to call this out in the docs for `ReadAll`. Otherwise, best to keep it simple

Reply:
There is, in fact, an explicit use case here. This function is more or less a drop-in replacement for [`io.ReadAll()`](https://pkg.go.dev/io#ReadAll) and it behaves the same way.

I have a use case where I want the data even if there is an error. If this function was to not return the read data on error, I would not be able to use it.

Put differently, getting both data and an error gives the caller more flexibility vs just getting one of them. gRPC may need this too one day.

The requirement to free the buffers is already documented on the function - line 236.